[stdlib_linalg] matrix property checks (#499)
fortran-lang/stdlib: doc/specs/stdlib_linalg.md

    !A = reshape([3., 6., 9., 4., 8., 12.], [3,2])
end program demo_outer_product
```

## `is_square` - Checks if a matrix is square

### Status

Experimental

### Description

Checks if a matrix is square

### Syntax

> `d = [[stdlib_linalg(module):is_square(interface)]](A)`
> 
> ### Arguments
> 
> `A`: Shall be a rank-2 array

Reviewer: instead of A, I would suggest to use the term `array`, which is more explicit to me.
```suggestion
`d = [[stdlib_linalg(module):is_square(interface)]](array)`

### Arguments

`array`: Shall be a rank-2 array
```

Author: I'm not a huge fan of this change since I think the documentation is already sufficiently (and exceedingly) clear, and having a variable (even an example variable) named after a built in structure seems like a bad idea.
I could very well be the odd one out on this though, so if other people like the proposed change I'm happy to make it.

Reviewer: Ok. Keep it like it is now.

### Return value

> Returns a logical value that is true if the input matrix is square, and false otherwise.

Reviewer: ```suggestion
Returns a `logical` scalar that is `.true.` if the input matrix is square, and `.false.` otherwise.
```

Author: Thanks for the suggestion, these changes have been made.

### Example

```fortran
program demo_is_square
    use stdlib_linalg, only: is_square
    implicit none
    real :: A(2,2), B(3,2)
    logical :: res
    A = reshape([1., 2., 3., 4.], shape(A))
    B = reshape([1., 2., 3., 4., 5., 6.], shape(B))
    res = is_square(A) ! returns .true.
    res = is_square(B) ! returns .false.
end program demo_is_square
```

## `is_diagonal` - Checks if a matrix is diagonal

### Status

Experimental

### Description

Checks if a matrix is diagonal

### Syntax

> `d = [[stdlib_linalg(module):is_diagonal(interface)]](A)`
> 
> ### Arguments
> 
> `A`: Shall be a rank-2 array

Reviewer: ```suggestion
`d = [[stdlib_linalg(module):is_diagonal(interface)]](array)`

### Arguments

`array`: Shall be a rank-2 array
```

### Return value

> Returns a logical value that is true if the input matrix is diagonal, and false otherwise.

Reviewer: ```suggestion
Returns a `logical` scalar that is `.true.` if the input matrix is diagonal, and `.false.` otherwise.
```

Note that nonsquare matrices may be diagonal, so long as `a_ij = 0` when `i /= j`.

### Example

```fortran
program demo_is_diagonal
    use stdlib_linalg, only: is_diagonal
    implicit none
    real :: A(2,2), B(2,2)
    logical :: res
    A = reshape([1., 0., 0., 4.], shape(A))
    B = reshape([1., 0., 3., 4.], shape(B))
    res = is_diagonal(A) ! returns .true.
    res = is_diagonal(B) ! returns .false.
end program demo_is_diagonal
```

## `is_symmetric` - Checks if a matrix is symmetric

### Status

Experimental

### Description

Checks if a matrix is symmetric

### Syntax

> `d = [[stdlib_linalg(module):is_symmetric(interface)]](A)`
> 
> ### Arguments
> 
> `A`: Shall be a rank-2 array

Reviewer: ```suggestion
`d = [[stdlib_linalg(module):is_symmetric(interface)]](array)`

### Arguments

`array`: Shall be a rank-2 array
```

### Return value

> Returns a logical value that is true if the input matrix is symmetric, and false otherwise.

Reviewer: ```suggestion
Returns a `logical` scalar that is `.true.` if the input matrix is symmetric, and `.false.` otherwise.
```

### Example

```fortran
program demo_is_symmetric
    use stdlib_linalg, only: is_symmetric
    implicit none
    real :: A(2,2), B(2,2)
    logical :: res
    A = reshape([1., 3., 3., 4.], shape(A))
    B = reshape([1., 0., 3., 4.], shape(B))
    res = is_symmetric(A) ! returns .true.
    res = is_symmetric(B) ! returns .false.
end program demo_is_symmetric
```

## `is_skew_symmetric` - Checks if a matrix is skew-symmetric

### Status

Experimental

### Description

Checks if a matrix is skew-symmetric

### Syntax

> `d = [[stdlib_linalg(module):is_skew_symmetric(interface)]](A)`
> 
> ### Arguments
> 
> `A`: Shall be a rank-2 array

Reviewer: ```suggestion
`d = [[stdlib_linalg(module):is_skew_symmetric(interface)]](A)`

### Arguments

`A`: Shall be a rank-2 array
```
```suggestion
`d = [[stdlib_linalg(module):is_skew_symmetric(interface)]](array)`

### Arguments

`array`: Shall be a rank-2 array
```

### Return value

> Returns a logical value that is true if the input matrix is skew-symmetric, and false otherwise.

Reviewer: ```suggestion
Returns a `logical` scalar that is `.true.` if the input matrix is skew-symmetric, and `.false.` otherwise.
```

### Example

```fortran
program demo_is_skew_symmetric
    use stdlib_linalg, only: is_skew_symmetric
    implicit none
    real :: A(2,2), B(2,2)
    logical :: res
    A = reshape([0., -3., 3., 0.], shape(A))
    B = reshape([0., 3., 3., 0.], shape(B))
    res = is_skew_symmetric(A) ! returns .true.
    res = is_skew_symmetric(B) ! returns .false.
end program demo_is_skew_symmetric
```

## `is_hermitian` - Checks if a matrix is Hermitian

### Status

Experimental

### Description

Checks if a matrix is Hermitian

### Syntax

> `d = [[stdlib_linalg(module):is_hermitian(interface)]](A)`
> 
> ### Arguments
> 
> `A`: Shall be a rank-2 array

Reviewer: ```suggestion
`d = [[stdlib_linalg(module):is_hermitian(interface)]](array)`

### Arguments

`array`: Shall be a rank-2 array
```

### Return value

> Returns a logical value that is true if the input matrix is Hermitian, and false otherwise.

Reviewer: ```suggestion
Returns a `logical` scalar that is `.true.` if the input matrix is Hermitian, and `.false.` otherwise.
```

### Example

```fortran
program demo_is_hermitian
    use stdlib_linalg, only: is_hermitian
    implicit none
    complex :: A(2,2), B(2,2)
    logical :: res
    A = reshape([cmplx(1.,0.), cmplx(3.,-1.), cmplx(3.,1.), cmplx(4.,0.)], shape(A))
    B = reshape([cmplx(1.,0.), cmplx(3.,1.), cmplx(3.,1.), cmplx(4.,0.)], shape(B))
    res = is_hermitian(A) ! returns .true.
    res = is_hermitian(B) ! returns .false.
end program demo_is_hermitian
```

## `is_triangular` - Checks if a matrix is triangular

### Status

Experimental

### Description

Checks if a matrix is triangular

### Syntax

> `d = [[stdlib_linalg(module):is_triangular(interface)]](A,uplo)`
> 
> ### Arguments
> 
> `A`: Shall be a rank-2 array

Reviewer: ```suggestion
`d = [[stdlib_linalg(module):is_triangular(interface)]](array,uplo)`

### Arguments

`array`: Shall be a rank-2 array
```

`uplo`: Shall be a single character from `{'u','U','l','L'}`

### Return value

> Returns a logical value that is true if the input matrix is the type of triangular specified by `uplo` (upper or lower), and false otherwise.

Reviewer: ```suggestion
Returns a `logical` scalar that is `.true.` if the input matrix is the type of triangular specified by `uplo` (upper or lower), and `.false.` otherwise.
```

> Note that the definition of triangular used here allows nonsquare matrices to be triangular.

Reviewer: ```suggestion
Note that the definition of triangular used in this implementation allows nonsquare matrices to be triangular.
```

Author: I agree, this is better. Changes have been made.

Specifically, upper triangular matrices satisfy `a_ij = 0` when `j < i`, and lower triangular matrices satisfy `a_ij = 0` when `j > i`.

### Example

```fortran
program demo_is_triangular
    use stdlib_linalg, only: is_triangular
    implicit none
    real :: A(3,3), B(3,3)
    logical :: res
    A = reshape([1., 0., 0., 4., 5., 0., 7., 8., 9.], shape(A))
    B = reshape([1., 0., 3., 4., 5., 0., 7., 8., 9.], shape(B))
    res = is_triangular(A,'u') ! returns .true.
    res = is_triangular(B,'u') ! returns .false.
end program demo_is_triangular
```

## `is_hessenberg` - Checks if a matrix is hessenberg

### Status

Experimental

### Description

Checks if a matrix is Hessenberg

### Syntax

> `d = [[stdlib_linalg(module):is_hessenberg(interface)]](A,uplo)`
> 
> ### Arguments
> 
> `A`: Shall be a rank-2 array

Reviewer: ```suggestion
`d = [[stdlib_linalg(module):is_hessenberg(interface)]](array,uplo)`

### Arguments

`array`: Shall be a rank-2 array
```

`uplo`: Shall be a single character from `{'u','U','l','L'}`

### Return value

> Returns a logical value that is true if the input matrix is the type of Hessenberg specified by `uplo` (upper or lower), and false otherwise.

Reviewer: ```suggestion
Returns a `logical` scalar that is `.true.` if the input matrix is the type of Hessenberg specified by `uplo` (upper or lower), and `.false.` otherwise.
```

> Note that the definition of Hessenberg used here allows nonsquare matrices to be Hessenberg.

Reviewer: ```suggestion
Note that the definition of Hessenberg used in this implementation allows nonsquare matrices to be Hessenberg.
```

### Status

Experimental

### Description

Checks if a matrix is diagonal

### Syntax

`d = [[stdlib_linalg(module):is_diagonal(interface)]](A)`

### Arguments

`A`: Shall be a rank-2 array

### Return value

Returns a logical value that is true if the input matrix is diagonal, and false otherwise.
Note that nonsquare matrices may be diagonal, so long as `a_ij = 0` when `i /= j`.

### Example

```fortran
program demo_is_diagonal
    use stdlib_linalg, only: is_diagonal
    implicit none
    real :: A_true(2,2), A_false(2,2)
    logical :: res
    A_true = reshape([1., 0., 0., 4.], shape(A_true))
    A_false = reshape([1., 0., 3., 4.], shape(A_false))
    res = is_diagonal(A_true)
    !res = .true.
    res = is_diagonal(A_false)
    !res = .false.
end program demo_is_diagonal
```

## `is_symmetric` - Checks if a matrix is symmetric

### Status

Experimental

### Description

Checks if a matrix is symmetric

### Syntax

`d = [[stdlib_linalg(module):is_symmetric(interface)]](A)`

### Arguments

`A`: Shall be a rank-2 array

### Return value

Returns a logical value that is true if the input matrix is symmetric, and false otherwise.

### Example

```fortran

> program demo_is_symmetric
>     use stdlib_linalg, only: is_symmetric
>     implicit none
>     real :: A_true(2,2), A_false(2,2)
>     logical :: res
>     A_true = reshape([1., 3., 3., 4.], shape(A_true))
>     A_false = reshape([1., 0., 3., 4.], shape(A_false))
>     res = is_symmetric(A_true)
>     !res = .true.
>     res = is_symmetric(A_false)
>     !res = .false.
> end program demo_is_symmetric

Reviewer 2: ```suggestion
program demo_is_symmetric
    use stdlib_linalg, only: is_symmetric
    implicit none
    real :: A(2,2), B(2,2)
    logical :: res
    A = reshape([1., 3., 3., 4.], shape(A))
    B = reshape([1., 0., 3., 4.], shape(B))
    res = is_symmetric(A) ! returns .true.
    res = is_symmetric(B) ! returns .false.
end program demo_is_symmetric
```

In light of the discussion at https://github.com/fortran-lang/stdlib/pull/539#discussion_r717399771, one more nitpick would be to revise the demonstration examples. 

@ghbrown, @milancurcic How do you feel about the clarity of my proposed patch?

Reviewer 3: I like it!

Author: I don't have a strong opinion on this one, but since other people seem to like the `return` version better I'm happy to make the change.
I'll go ahead and edit the other tests as well.

Author: All examples should be updated, thanks for the feedback.
